File: frontend/js/paymentMethods.js
const form = document.getElementById('paymentMethodForm');
const paymentType = document.getElementById('paymentType');
const paymentMethodId = document.getElementById('paymentMethodId');
const submitPaymentButton = document.getElementById('submitPaymentButton');
const cancelEditButton = document.getElementById('cancelEditButton');
const searchPaymentType = document.getElementById('searchPaymentType');
const paymentMethodList = document.getElementById('paymentMethodList');
const emptyPaymentMessage = document.getElementById('emptyPaymentMessage');
const addPaymentButton = document.getElementById('addPaymentButton');
const paymentModal = document.getElementById('paymentModal');
const closePaymentModal = document.getElementById('closePaymentModal');
const deleteConfirmModal = document.getElementById('deleteConfirmModal');
const confirmDeleteButton = document.getElementById('confirmDeleteButton');
const cancelDeleteButton = document.getElementById('cancelDeleteButton');

const API_BASE_URL = 'http://127.0.0.1:8080';

let paymentMethods = [];
let paymentMethodIdToDelete = '';
let canManagePaymentMethods = true;

function getPaymentHeaders() {
    return {
        'Content-Type': 'application/json'
    };
}

function getStoredUserId() {
    return sessionStorage.getItem('userId');
}

function withStoredUserId(path) {
    const userId = getStoredUserId();

    if (!userId) {
        return path;
    }

    const separator = path.includes('?') ? '&' : '?';
    return `${path}${separator}user_id=${encodeURIComponent(userId)}`;
}

function checkCustomerControls() {
    const role = sessionStorage.getItem('userRole');
    if (role !== 'customer') {
        addPaymentButton.style.display = 'none';
    } else {
        addPaymentButton.style.display = 'block';
    }
}

function showApiError(message) {
    alert(message || 'Something went wrong. Please try again.');
}

async function requestPaymentMethods(path, options) {
    const response = await fetch(`${API_BASE_URL}${withStoredUserId(path)}`, {
        credentials: 'include',
        ...options,
        headers: {
            ...getPaymentHeaders(),
            ...(options && options.headers ? options.headers : {})
        }
    });

    if (response.status === 204) {
        return null;
    }

    const data = await response.json();

    if (!response.ok) {
        const error = new Error(data.error || 'Payment method request failed');
        error.errors = data.errors;
        throw error;
    }

    return data;
}

async function loadPaymentMethods() {
    if (!getStoredUserId()) {
        window.location.href = 'login.html';
        return;
    }

    try {
        paymentMethods = await requestPaymentMethods('/payment-methods/');
        canManagePaymentMethods = true;
        searchPaymentType.disabled = false;
        addPaymentButton.disabled = false;
        renderPaymentMethods();
    } catch (error) {
        canManagePaymentMethods = false;
        paymentMethods = [];
        paymentMethodList.innerHTML = '';
        searchPaymentType.disabled = true;
        addPaymentButton.disabled = true;
        emptyPaymentMessage.textContent = error.message;
        emptyPaymentMessage.classList.add('is-error');
        emptyPaymentMessage.style.display = 'block';
    }
}

function showPaymentForm() {
    if (!canManagePaymentMethods) {
        return;
    }

    paymentModal.classList.remove('hidden');
    document.body.classList.add('modal-open');
    paymentType.focus();
}

function hidePaymentForm() {
    paymentModal.classList.add('hidden');
    document.body.classList.remove('modal-open');
}

function showDeleteConfirm(methodId) {
    if (!canManagePaymentMethods) {
        return;
    }

    paymentMethodIdToDelete = methodId;
    deleteConfirmModal.classList.remove('hidden');
    document.body.classList.add('modal-open');
    confirmDeleteButton.focus();
}

function hideDeleteConfirm() {
    paymentMethodIdToDelete = '';
    deleteConfirmModal.classList.add('hidden');
    document.body.classList.remove('modal-open');
}

// delete the selected payment method after the confirmation popup is confirmed
async function deleteSelectedPaymentMethod() {
    if (paymentMethodIdToDelete === '') {
        return;
    }

    try {
        await requestPaymentMethods(getPaymentMethodEndpoint(paymentMethodIdToDelete), {
            method: 'DELETE'
        });
        hideDeleteConfirm();
        await loadPaymentMethods();
    } catch (error) {
        showApiError(error.message);
    }
}

// shows only the form fields that match the selected payment method type
function showFieldsForSelectedType() {
    document.querySelectorAll('[data-payment-fields]').forEach(function (fieldGroup) {
        fieldGroup.classList.toggle('hidden', fieldGroup.dataset.paymentFields !== paymentType.value);
    });
    clearErrors();
}

function clearErrors() {
    document.querySelectorAll('.error').forEach(function (error) {
        error.textContent = '';
        error.style.display = 'none';
    });

    document.querySelectorAll('input, select').forEach(function (field) {
        field.classList.remove('invalid');
    });
}

function showError(fieldId, message) {
    const field = document.getElementById(fieldId);
    const error = document.getElementById(`${fieldId}Error`);

    if (field) {
        field.classList.add('invalid');
    }

    if (error) {
        error.textContent = message;
        error.style.display = 'block';
    }
}

function showBackendErrors(errors) {
    Object.keys(errors || {}).forEach(function (fieldId) {
        showError(fieldId, errors[fieldId]);
    });
}

function getNumbersOnly(value) {
    return value.replace(/\D/g, ''); // \D not a digit, g globally
}

function isNumbersOnly(value) {
    return /^\d+$/.test(value); // \d+ means one or more digits
}

function isLettersOnly(value) {
    return /^[A-Za-z ]+$/.test(value); 
}

function isValidExpiry(expiryDate) {
    const match = expiryDate.match(/^(0[1-9]|1[0-2])\/(\d{2})$/); // MM/YY

    if (!match) {
        return false;
    }

    const month = Number(match[1]);
    const year = Number(`20${match[2]}`);
    const lastDayOfExpiryMonth = new Date(year, month, 0);
    const today = new Date();

    today.setHours(0, 0, 0, 0);
    return lastDayOfExpiryMonth >= today;
}

function getNextPaymentMethodId() {
    const highestId = paymentMethods.reduce(function (highest, method) {
        const numericId = Number(method.id);
        return Number.isNaN(numericId) ? highest : Math.max(highest, numericId);
    }, 0);

    return String(highestId + 1).padStart(3, '0');
}

// checks all visible payment method fields before allowing the method is added
function validatePaymentMethod() {
    clearErrors();

    const type = paymentType.value;
    const existingMethod = paymentMethods.find(function (method) {
        return method.id === paymentMethodId.value;
    });
    let isValid = true;

    if (type === 'card') {
        const cardholderName = document.getElementById('cardholderName').value.trim();
        const cardNumber = document.getElementById('cardNumber').value.trim();
        const expiryDate = document.getElementById('expiryDate').value.trim();
        const cvv = document.getElementById('cvv').value.trim();
        const isUsingSavedCardNumber = existingMethod && cardNumber.includes('*');

        if (cardholderName === '') {
            showError('cardholderName', 'Cardholder name is required');
            isValid = false;
        } else if (!isLettersOnly(cardholderName)) {
            showError('cardholderName', 'Cardholder name can only contain letters');
            isValid = false;
        }

        if (!isUsingSavedCardNumber && !isNumbersOnly(cardNumber)) {
            showError('cardNumber', 'Card number can only contain numbers');
            isValid = false;
        }

        if (!isValidExpiry(expiryDate)) {
            showError('expiryDate', 'Use a valid future expiry date');
            isValid = false;
        }

        if (!existingMethod && !isNumbersOnly(cvv)) {
            showError('cvv', 'CVV can only contain numbers');
            isValid = false;
        }
    }

    if (type === 'bank') {
        const accountName = document.getElementById('accountName').value.trim();
        const bsbValue = document.getElementById('bsb').value;
        const accountNumberValue = document.getElementById('accountNumber').value;
        const isUsingSavedBsb = existingMethod && bsbValue.includes('*');
        const isUsingSavedAccountNumber = existingMethod && accountNumberValue.includes('*');

        if (accountName === '') {
            showError('accountName', 'Account name is required');
            isValid = false;
        } else if (!isLettersOnly(accountName)) {
            showError('accountName', 'Account name can only contain letters');
            isValid = false;
        }

        if (!isUsingSavedBsb && !isNumbersOnly(bsbValue)) {
            showError('bsb', 'BSB can only contain numbers');
            isValid = false;
        }

        if (!isUsingSavedAccountNumber && !isNumbersOnly(accountNumberValue)) {
            showError('accountNumber', 'Account number can only contain numbers');
            isValid = false;
        }
    }

    if (type === 'paypal') {
        const paypalEmail = document.getElementById('paypalEmail').value.trim();
        const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

        if (!existingMethod && !emailPattern.test(paypalEmail)) {
            showError('paypalEmail', 'Enter a valid PayPal email');
            isValid = false;
        }

        if (existingMethod && paypalEmail !== '' && !emailPattern.test(paypalEmail)) {
            showError('paypalEmail', 'Enter a valid PayPal email');
            isValid = false;
        }
    }

    return isValid;
}

// builds a payment method object from the form values, keeping sensitive details masked
function getPaymentMethodFromForm() {
    const type = paymentType.value;
    const existingMethod = paymentMethods.find(function (method) {
        return method.id === paymentMethodId.value;
    });
    const method = {
        id: paymentMethodId.value || getNextPaymentMethodId(),
        type,
        nickname: document.getElementById('nickname').value,
        createdAt: new Date().toISOString()
    };

    if (type === 'card') {
        const cardNumberValue = document.getElementById('cardNumber').value;
        method.cardholderName = document.getElementById('cardholderName').value.trim();
        method.cardNumber = cardNumberValue.includes('*') && existingMethod ? '' : cardNumberValue;
        method.expiryDate = document.getElementById('expiryDate').value.trim();
    }

    if (type === 'bank') {
        const bsbValue = document.getElementById('bsb').value;
        const accountNumberValue = document.getElementById('accountNumber').value;
        method.accountName = document.getElementById('accountName').value.trim();
        method.bsb = bsbValue.includes('*') && existingMethod ? '' : bsbValue;
        method.accountNumber = accountNumberValue.includes('*') && existingMethod ? '' : accountNumberValue;
    }

    if (type === 'paypal') {
        const paypalEmail = document.getElementById('paypalEmail').value.trim();
        method.paypalEmail = paypalEmail;
    }

    return method;
}

function getCardBrand(cardNumber) {
    if (/^4/.test(cardNumber)) {
        return 'Visa';
    }

    if (/^5[1-5]/.test(cardNumber)) {
        return 'Mastercard';
    }

    if (/^3[47]/.test(cardNumber)) {
        return 'American Express';
    }

    return 'Card';
}

function maskEmail(email) {
    const parts = email.split('@');
    const name = parts[0];
    const domain = parts[1];
    const visibleName = name.length <= 2 ? name.charAt(0) : name.slice(0, 2);

    return `${visibleName}***@${domain}`;
}

// clears the form and returns it to the default add payment state
function resetForm() {
    form.reset();
    paymentMethodId.value = '';
    paymentType.value = 'card';
    submitPaymentButton.textContent = 'Add Payment Method';
    cancelEditButton.textContent = 'Cancel';
    cancelEditButton.hidden = false;
    showFieldsForSelectedType();
}

function getTypeLabel(type) {
    const labels = {
        card: 'Credit or Debit Card',
        bank: 'Bank Account Transfer',
        paypal: 'PayPal'
    };

    return labels[type] || type;
}

function getMethodDetail(method) {
    if (method.type === 'card') {
        return `${method.cardBrand || 'Card'} ending in ${method.cardLastFour} - expires ${method.expiryDate}`;
    }

    if (method.type === 'bank') {
        return `BSB ending in ${method.bsbLastThree} - account ending in ${method.accountLastFour}`;
    }

    return method.paypalEmail;
}

// rebuilds the payment method list based on the selected search type
function renderPaymentMethods() {
    const selectedType = searchPaymentType.value;
    const filteredMethods = selectedType === 'all'
        ? paymentMethods
        : paymentMethods.filter(function (method) {
        return method.type === selectedType;
        });

    paymentMethodList.innerHTML = '';
    emptyPaymentMessage.textContent = 'No payment methods saved yet.';
    emptyPaymentMessage.classList.remove('is-error');
    emptyPaymentMessage.style.display = filteredMethods.length === 0 ? 'block' : 'none';

    filteredMethods.forEach(function (method) {
        const methodCard = document.createElement('article');
        const methodDetails = document.createElement('div');
        const typeLabel = document.createElement('span');
        const nickname = document.createElement('h3');
        const detail = document.createElement('p');
        const actions = document.createElement('div');
        const editButton = document.createElement('button');
        const deleteButton = document.createElement('button');

        methodCard.className = 'payment-method-card';
        typeLabel.className = 'payment-method-type';
        actions.className = 'payment-card-actions';
        editButton.className = 'secondary-payment-button';
        deleteButton.className = 'danger-payment-button';

        typeLabel.textContent = getTypeLabel(method.type);
        nickname.textContent = method.nickname;
        detail.textContent = getMethodDetail(method);

        editButton.type = 'button';
        editButton.dataset.action = 'edit';
        editButton.dataset.id = method.id;
        editButton.textContent = 'Edit';

        deleteButton.type = 'button';
        deleteButton.dataset.action = 'delete';
        deleteButton.dataset.id = method.id;
        deleteButton.textContent = 'Delete';

        methodDetails.append(typeLabel, nickname, detail);
        actions.append(editButton, deleteButton);
        methodCard.append(methodDetails, actions);

        paymentMethodList.appendChild(methodCard);
    });
}

// opens the popup and fill the form with the existing payment method for editing
function fillFormForEdit(method) {
    resetForm();
    showPaymentForm();

    paymentMethodId.value = method.id;
    paymentType.value = method.type;
    document.getElementById('nickname').value = method.nickname;
    showFieldsForSelectedType();

    if (method.type === 'card') {
        document.getElementById('cardholderName').value = method.cardholderName || '';
        document.getElementById('cardNumber').value = method.cardLastFour ? `************${method.cardLastFour}` : '';
        document.getElementById('expiryDate').value = method.expiryDate || '';
        document.getElementById('cvv').value = '';
    }

    if (method.type === 'bank') {
        document.getElementById('accountName').value = method.accountName || '';
        document.getElementById('bsb').value = method.bsbLastThree ? `***${method.bsbLastThree}` : '';
        document.getElementById('accountNumber').value = method.accountLastFour ? `******${method.accountLastFour}` : '';
    }

    if (method.type === 'paypal') {
        document.getElementById('paypalEmail').value = '';
    }

    submitPaymentButton.textContent = 'Update Payment Method';
    cancelEditButton.textContent = 'Cancel Edit';
    cancelEditButton.hidden = false;
}

// replaces the saved payment method with the updated form values
function getPaymentMethodEndpoint(methodId) {
    return `/payment-methods/${Number(methodId)}`;
}

form.addEventListener('submit', async function (event) {
    event.preventDefault();

    if (!validatePaymentMethod()) {
        return;
    }

    const method = getPaymentMethodFromForm();

    try {
        if (paymentMethodId.value) {
            await requestPaymentMethods(getPaymentMethodEndpoint(paymentMethodId.value), {
                method: 'PUT',
                body: JSON.stringify(method)
            });
        } else {
            await requestPaymentMethods('/payment-methods/', {
                method: 'POST',
                body: JSON.stringify(method)
            });
        }

        await loadPaymentMethods();
        resetForm();
        hidePaymentForm();
    } catch (error) {
        if (error.errors) {
            showBackendErrors(error.errors);
            return;
        }

        showApiError(error.message);
    }
});

paymentType.addEventListener('change', showFieldsForSelectedType);

paymentMethodList.addEventListener('click', function (event) {
    const button = event.target.closest('button[data-action]');

    if (!button) {
        return;
    }

    const methodId = button.dataset.id;
    const method = paymentMethods.find(function (paymentMethod) {
        return paymentMethod.id === methodId;
    });

    if (!method) {
        return;
    }

    if (button.dataset.action === 'edit') {
        fillFormForEdit(method);
        return;
    }

    showDeleteConfirm(methodId);
});

searchPaymentType.addEventListener('change', renderPaymentMethods);

cancelEditButton.addEventListener('click', function () {
    resetForm();
    hidePaymentForm();
});

closePaymentModal.addEventListener('click', function () {
    resetForm();
    hidePaymentForm();
});

paymentModal.addEventListener('click', function (event) {
    if (event.target === paymentModal) {
        resetForm();
        hidePaymentForm();
    }
});

confirmDeleteButton.addEventListener('click', deleteSelectedPaymentMethod);

cancelDeleteButton.addEventListener('click', hideDeleteConfirm);

deleteConfirmModal.addEventListener('click', function (event) {
    if (event.target === deleteConfirmModal) {
        hideDeleteConfirm();
    }
});

addPaymentButton.addEventListener('click', function () {
    resetForm();
    showPaymentForm();
});

checkCustomerControls();
showFieldsForSelectedType();
loadPaymentMethods();

File: backend/routes/payment_methods.py
from datetime import date
from flask import Blueprint, request, jsonify, session

from backend.models.payment_method_crud import (
    create_payment_method,
    delete_payment_method,
    get_payment_method,
    is_customer,
    list_payment_methods,
    update_payment_method,
)

payment_methods_bp = Blueprint("payment_methods", __name__, url_prefix="/payment-methods")

VALID_TYPES = {"card", "bank", "paypal"}


def numbers_only(value):
    return "".join(character for character in (value or "") if character.isdigit())


def get_card_brand(card_number):
    if card_number.startswith("4"):
        return "Visa"
    if len(card_number) >= 2 and card_number[0] == "5" and card_number[1] in "12345":
        return "Mastercard"
    if len(card_number) >= 2 and card_number[0] == "3" and card_number[1] in "47":
        return "American Express"
    return "Card"


def mask_email(email):
    name, domain = email.split("@", 1)
    visible_name = name[0] if len(name) <= 2 else name[:2]
    return f"{visible_name}***@{domain}"


def is_valid_expiry(expiry_date):
    if not expiry_date or len(expiry_date) != 5 or expiry_date[2] != "/":
        return False

    month_text, year_text = expiry_date.split("/")

    if not month_text.isdigit() or not year_text.isdigit():
        return False

    month = int(month_text)
    if month < 1 or month > 12:
        return False

    year = int(f"20{year_text}")
    today = date.today()
    return year > today.year or (year == today.year and month >= today.month)


def is_valid_email(email):
    if not email or email.count("@") != 1:
        return False

    name, domain = email.split("@", 1)
    if not name or not domain or " " in email:
        return False

    if "." not in domain:
        return False

    domain_parts = domain.split(".")
    return all(domain_parts)


def current_customer_id():
    user_id = session.get("user_id")
    if user_id:
        return user_id if is_customer(user_id) else None

    stored_user_id = request.args.get("user_id")
    if not stored_user_id:
        return None

    try:
        user_id = int(stored_user_id)
    except ValueError:
        return None

    return user_id if is_customer(user_id) else None


def require_customer():
    user_id = current_customer_id()
    if not user_id:
        if session.get("user_id") or request.args.get("user_id"):
            return None, (jsonify({"error": "Only customers can manage payment methods"}), 403)
        return None, (jsonify({"error": "Login is required"}), 401)

    return user_id, None


def parse_payment_method(payload, existing_method=None):
    errors = {}
    method_type = (payload.get("type") or "").strip().lower()
    nickname = payload.get("nickname") or ""
    data = {
        "type": method_type,
        "nickname": nickname,
        "cardholder_name": None,
        "card_brand": None,
        "card_last_four": None,
        "expiry_date": None,
        "account_name": None,
        "bsb_last_three": None,
        "account_last_four": None,
        "paypal_email_masked": None,
    }

    if method_type not in VALID_TYPES:
        errors["paymentType"] = "Choose a valid payment type"

    if method_type == "card":
        cardholder_name = (payload.get("cardholderName") or "").strip()
        card_number = numbers_only(payload.get("cardNumber"))
        expiry_date = (payload.get("expiryDate") or "").strip()

        if not cardholder_name:
            errors["cardholderName"] = "Cardholder name is required"

        if not existing_method and len(card_number) < 12:
            errors["cardNumber"] = "Enter a valid card number"

        if existing_method and not card_number:
            data["card_last_four"] = existing_method.card_last_four
            data["card_brand"] = existing_method.card_brand
        elif card_number:
            data["card_last_four"] = card_number[-4:]
            data["card_brand"] = get_card_brand(card_number)

        if not is_valid_expiry(expiry_date):
            errors["expiryDate"] = "Use a valid future expiry date"

        data["cardholder_name"] = cardholder_name
        data["expiry_date"] = expiry_date

    if method_type == "bank":
        account_name = (payload.get("accountName") or "").strip()
        bsb = numbers_only(payload.get("bsb"))
        account_number = numbers_only(payload.get("accountNumber"))

        if not account_name:
            errors["accountName"] = "Account name is required"

        if not existing_method and len(bsb) != 6:
            errors["bsb"] = "BSB must be 6 digits"

        if not existing_method and len(account_number) < 6:
            errors["accountNumber"] = "Enter a valid account number"

        if existing_method and not bsb:
            data["bsb_last_three"] = existing_method.bsb_last_three
        elif bsb:
            data["bsb_last_three"] = bsb[-3:]

        if existing_method and not account_number:
            data["account_last_four"] = existing_method.account_last_four
        elif account_number:
            data["account_last_four"] = account_number[-4:]

        data["account_name"] = account_name

    if method_type == "paypal":
        paypal_email = (payload.get("paypalEmail") or "").strip()

        if existing_method and not paypal_email:
            data["paypal_email_masked"] = existing_method.paypal_email_masked
        elif is_valid_email(paypal_email):
            data["paypal_email_masked"] = mask_email(paypal_email)
        else:
            errors["paypalEmail"] = "Enter a valid PayPal email"

    return data, errors


@payment_methods_bp.route("/", methods=["GET"])
def get_payment_methods():
    user_id, error_response = require_customer()
    if error_response:
        return error_response

    method_type = request.args.get("type")
    if method_type and method_type not in VALID_TYPES:
        return jsonify({"error": "Invalid payment type"}), 400

    methods = list_payment_methods(user_id, method_type)
    return jsonify([method.to_dict() for method in methods])


@payment_methods_bp.route("/", methods=["POST"])
def add_payment_method():
    user_id, error_response = require_customer()
    if error_response:
        return error_response

    method_data, errors = parse_payment_method(request.get_json(silent=True) or {})
    if errors:
        return jsonify({"errors": errors}), 400

    method = create_payment_method(user_id, method_data)
    return jsonify(method.to_dict()), 201


@payment_methods_bp.route("/<int:payment_method_id>", methods=["GET"])
def get_single_payment_method(payment_method_id):
    user_id, error_response = require_customer()
    if error_response:
        return error_response

    method = get_payment_method(user_id, payment_method_id)
    if not method:
        return jsonify({"error": "Payment method not found"}), 404

    return jsonify(method.to_dict())


@payment_methods_bp.route("/<int:payment_method_id>", methods=["PUT"])
def edit_payment_method(payment_method_id):
    user_id, error_response = require_customer()
    if error_response:
        return error_response

    existing_method = get_payment_method(user_id, payment_method_id)
    if not existing_method:
        return jsonify({"error": "Payment method not found"}), 404

    method_data, errors = parse_payment_method(
        request.get_json(silent=True) or {},
        existing_method=existing_method,
    )
    if errors:
        return jsonify({"errors": errors}), 400

    method = update_payment_method(user_id, payment_method_id, method_data)
    return jsonify(method.to_dict())


@payment_methods_bp.route("/<int:payment_method_id>", methods=["DELETE"])
def remove_payment_method(payment_method_id):
    user_id, error_response = require_customer()
    if error_response:
        return error_response

    if not delete_payment_method(user_id, payment_method_id):
        return jsonify({"error": "Payment method not found"}), 404

    return "", 204
